### ISSUE TITLE
ENH: De-couple `Data` from `Survey`

### ISSUE THREAD

**Issue by prisae** [simpeg/simpeg#1530]
### Proposed new feature or change:

# De-couple `Data` from `Survey`

## History

This is simply a brainstorm of things that started over 7.5 years ago at the [opengeophysics/scratchpad](https://github.com/opengeophysics/scratchpad). When I started with empymod, and later emg3d, I always wanted to focus on the forward modelling, and use existing frameworks for the inversion, not repeating that bit. Back then @lheagy and I started to brainstorm how it could work and what is missing.

@sgkang managed to get SimPEG(emg3d) working some 3+ years ago, which he described in the curvenote [emg3d-as-solver-for-simpeg](https://curvenote.com/@prisae/emg3d-as-solver-for-simpeg/hackathon-emg3d-inversion-in-simpeg).

I recently retook that work, together with other examples for

- SimPEG(emg3d)
- SimPEG(empymod)
- pyGIMLi(emg3d)
- pyGIMLi(empymod)

and collected working examples in [emsig/devInv](https://github.com/emsig/devInv) -- the emg3d inversion part will hopefully get into emg3d-main and released this year (development at [emsig/emg3d/tree/inv-simpeg/emg3d](https://github.com/emsig/emg3d/tree/inv-simpeg/emg3d)). And the empymod part follows hopefully next year.

## What is possible

When retaking this work I realized that in the meantime, almost everything is in place in SimPEG so that any forward modeller can use the framework for inversion. The basic required pieces are, in dummy code, the following, assuming `Fwd` is any forward modeller with certain attributes (`.Jvec`, `.Jtvec`, `.dpred`),

```python
class SomeSimulation(simpeg.SomeBaseSimulation):

    def __init__(self, Fwd, mesh, survey, **kwargs):
        self.f = Fwd
        # mesh: a discretize mesh
        # survey: a SimPEG survey
        super().__init__(mesh, survey, **kwargs)

    def Jvec(self, m, v, f=None):
        if f is None:
            f = self.fields(m=m)
        return f.Jvec(m, v)

    def Jtvec(self, m, v, f=None):
        if f is None:
            f = self.fields(m=m)
        return f.Jtvec(m, v)

    def dpred(self, m=None, f=None):
        if f is None:
            f = self.fields(m=m)
        return f.dpred(m, f)

    def fields(self, m=None):
        if m is not None:
            self.f.update_model(m=m)
        return self.f

data = simpeg.data.Data(survey, dobs, standard_deviation)
# with survey a SimPEG survey
``` 

This works pretty good and is what I use now.

## What could be improved

The only tricky bit, in my opinion, is the need for a SimPEG survey, and its entanglement with the SimPEG data container. As far as I could see, the SimPEG survey is, in my case, only used for `survey.nD` and `survey.vnD`. Of these, `nD` would already be available in `data.nD`. I create the survey with dummy-replacements (in my case `FDEM.sources.BaseFDEMSrc` and `FDEM.receivers.BaseRx`, where FDEM is `simpeg.electromagnetics.frequency_domain`, with any arbitrary location, orientation, dtype, etc, it does not matter. The only thing that matters, when I translate my emg3d survey to a simpeg survey, is the order of the data. The rest of the survey is irrelevant.

## Suggestion

So basically the suggestion would be to **De-tangle Data from Survey, and give the possibility for `Survey=None` when creating a Simulation or a Data container.**

I have no idea how easy or difficult it would be, if there is interest or not, or if something along these lines is already in the works. I just thought I jot down these ideas here. I think it would open up / simplify the possibility of using SimPEG as an inversion framework for any forward model.

**Comment by prisae** [comment]
If it is of interest: this is also how it works in pyGIMLi. The inversion frameworks requires a mesh, and functions which return the simulated data, Jvec (mult), and Jtvec (transMult); without the need for a survey.

**Comment by santisoler** [comment]
Thanks for opening this @prisae. I agree with the idea of detaching the inversion framework in SimPEG from the rest of the machinery in the package, so it could easily be used with any other type of forward model or to run other type of inversions (e.g.: parametric inversions might not require a mesh).

Ultimately, I think we should separate the bits related the optimization problem (objective functions, data misfits, regularizations, optimization classes, data, uncertainties, etc) from bits related to forwards (simulations, surveys, sources, receivers, etc). I think we should keep the interconnections between these two as a minimum, so if you want to use your own forward with the existing inversion framework you need to write little code, and vice versa (if you want to use your own inversion framework with one of simpeg's forwards).

To achieve this goal, we need to think how to lower the surface area between these two. For example, the forward models don't need to know about the observed data, and the data misfits don't need to know about the survey.

A first step towards this could be considering the de-entanglement you suggest. From my perspective, the `Data` class is a container for a data array and the uncertainties (standard deviation of the data), with some inner tools to get the uncertainties from the standard deviation, noise floor and/or relative error.

We could simplify this by making the data misfit class to take the data array and its corresponding standard deviation, and we could have an utility function to compute the uncertainties vector from the standard deviation, noise floor and/or relative error.

Something similar could be applied to the design of simulations. We could have a simple base class for simulations that don't assume that your forward requires a mesh or a survey. Such class should require users to implement a minimum set of methods that will be required to plug them into the data misfits and the rest of the inversion framework.

For example, in very simple code, we could make the data misfit classes to take the `data` and the `standard_deviation` (bypassing the need for a `Data` class):

```python
from numpy.typing import NDArray

class L2DataMisfit(BaseDataMisfit):

    def __init__(
        self, data: NDArray, standard_deviation: NDArray, simulation, **kwargs
    ): ...
```

And we could have a simple abstract class for all of our simulations (agnostic of surveys and meshes):

```python
from abc import ABC, abstractmethod

class BaseSimulation(ABC):

    @abstractmethod
    def Jvec(self, m, v, f=None):
        pass

    @abstractmethod
    def Jtvec(self, m, v, f=None):
        pass

    @abstractmethod
    def dpred(self, m=None, f=None):
        pass
```

So, if you want to plug your own forward to SimPEG's inversion framework, you could do something like:

```python
class MySimulation(BaseSimulation):
    """
    Implement my own forward model
    """

    def __init__(self, *args): ...

    def Jvec(self, m, v, f=None): ...

    def Jtvec(self, m, v, f=None): ...

    def dpred(self, m=None, f=None): ...


observed_data = ...  # read observed data from a file
standard_deviation = ...  # read observed data uncertainties from a file

simulation = MySimulation(...)
data_misfit = L2DataMisfit(observed_data, standard_deviation, simulation)
```

What do you think?

**Comment by prisae** [comment]
Yes, that sounds great and is more or less what I had in mind, but much more eloquently described! I just thought to jot down my experience. Feel free to close this one and open separate once for different/separate sub-issues, if it makes more sense.

**Comment by jcapriot** [comment]
Feel like Santi's comment is relevant for #1429

**Comment by jcapriot** [comment]
I still often use the `Data` container myself. It is very useful (but not well documented with examples) to index the data container using the survey's sources and receivers to retrieve the relevant values. It's also good to make sure that you have the data vector in the right order too.

**Comment by jcapriot** [comment]
I think it could be useful though to allow `survey=None` for the `Data` container.

**Comment by santisoler** [comment]
> I still often use the `Data` container myself. It is very useful (but not well documented with examples) to index the data container using the survey's sources and receivers to retrieve the relevant values. It's also good to make sure that you have the data vector in the right order too.

I see. I agree that having the `Data` class that allows us to index the data using the survey is neat. Considering that, I probably wouldn't just remove the class from SimPEG, but maybe consider it as a not necessary part to run the inversions. I think it would be simpler if you could just run an inversion passing a `data` and a `standard_deviation` arrays to the data misfit. We could also support the `data` argument to be a `Data` object, as we do now.
